nemo/collections/nlp/models/language_modeling/megatron/gpt_full_te_layer_autocast_spec.py

```diff
@@ -0,0 +1,273 @@
+        device: str = 'cuda',
+    ) -> None:
+        super().__init__(
+            hidden_size=hidden_size,
+            ffn_hidden_size=ffn_hidden_size,
+            layernorm_epsilon=layernorm_epsilon,
+            num_attention_heads=num_attention_heads,
+            init_method=init_method,
+            output_layer_init_method=output_layer_init_method,
+            hidden_dropout=hidden_dropout,
+            attention_dropout=attention_dropout,
+            layer_number=layer_number,
+            kv_channels=kv_channels,
+            self_attn_mask_type=self_attn_mask_type,
+            tp_group=tp_group,
+            tp_size=tp_size,
+            params_dtype=params_dtype,
+            get_rng_state_tracker=get_rng_state_tracker,
+            fuse_wgrad_accumulation=fuse_wgrad_accumulation,
+            seq_length=seq_length,
+            micro_batch_size=micro_batch_size,
+            sequence_parallel=sequence_parallel,
+            apply_residual_connection_post_layernorm=apply_residual_connection_post_layernorm,
+            output_layernorm=output_layernorm,
+            layer_type=layer_type,
+            drop_path_rate=drop_path_rate,
+            set_parallel_mode=tp_size > 1,
+            fuse_qkv_params=True,
+            zero_centered_gamma=zero_centered_gamma,
+            ub_tp_comm_overlap=ub_tp_comm_overlap,
+            device=device,
+        )
+        # use_emha=use_emha,
+
+        # Dtype for forward pass - ignore amp O2
+        self.dtype = utils_funcs.torch_dtype_from_precision(autocast_dtype, megatron_amp_O2=None)
+
+    def forward(
+        self,
+        hidden_states: torch.Tensor,
+        attention_mask: torch.Tensor,
+        encoder_output: Optional[torch.Tensor] = None,
+        enc_dec_attn_mask: Optional[torch.Tensor] = None,
+        inference_params: Optional[Any] = None,
+        is_first_microbatch: Optional[bool] = None,
+        checkpoint_core_attention: Optional[bool] = False,
+    ) -> torch.Tensor:
+        if self.dtype == torch.float32:
+            return super().forward(
+                hidden_states,
+                attention_mask,
+                encoder_output=encoder_output,
+                enc_dec_attn_mask=enc_dec_attn_mask,
+                inference_params=inference_params,
+                is_first_microbatch=is_first_microbatch,
+                checkpoint_core_attention=checkpoint_core_attention,
+            )
+        with torch.autocast(device_type="cuda", dtype=self.dtype):
+            return super().forward(
+                hidden_states,
+                attention_mask,
+                encoder_output=encoder_output,
+                enc_dec_attn_mask=enc_dec_attn_mask,
+                inference_params=inference_params,
+                is_first_microbatch=is_first_microbatch,
+                checkpoint_core_attention=checkpoint_core_attention,
+            )
+
+
+class TETransformerLayerAutocast(AutocastTransformerLayer):
```

Comment:
You might have to change this too

```suggestion
from megatron.core.transformer.transformer_layer import BaseTransformerLayer
class TETransformerLayerAutocast(AutocastTransformerLayer, BaseTransformerLayer):
```

Reply:
What is the purpose of this change?

Reply:
For megatron-core's `TransformerBlock` to recognize a different "TransformerLayer" (`TETransformerLayerAutocast`) in this case, it checks whether the new layer is a subclass of `BaseTransformerLayer`. So any new "TransformerLayer" wrapper will need to subclass it.

Reply:
We need this change as the MCore [PR](https://gitlab-master.nvidia.com/ADLR/megatron-lm/-/merge_requests/1113) has been merged.